core/mount/manager/manager_linux_test.go

```diff
@@ -0,0 +1,384 @@
+	}
+
+	if err := mount.All(ainfo.System, tm); err != nil {
+		t.Fatalf("failed to mount %v: %v", ainfo.System, err)
+	}
+	defer testutil.Unmount(t, tm)
+
+	if err := fstest.CheckDirectoryEqual(tm, actual); err != nil {
+		t.Fatalf("check directory failed: %v", err)
+	}
+}
+
+func TestLoopbackOverlay(t *testing.T) {
+	testutil.RequiresRoot(t)
+	ctx := logtest.WithT(context.Background(), t)
+	ctx = namespaces.WithNamespace(ctx, "test")
+	td := t.TempDir()
+	metadb := filepath.Join(td, "mounts.db")
+	targetdir := filepath.Join(td, "m")
+	db, err := bolt.Open(metadb, 0600, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	l1 := fstest.Apply(
+		fstest.CreateFile("/foo", []byte("foo\n"), 0777),
+		fstest.CreateDir("/a", 0755),
+		fstest.CreateDir("/a/b", 0755),
+		fstest.CreateDir("/a/b/c", 0755),
+	)
+
+	l2 := fstest.Apply(
+		fstest.CreateFile("/foo", []byte("foo\n"), 0777),
+		fstest.CreateDir("/a", 0755),
+		fstest.CreateDir("/a/b", 0755),
+		fstest.CreateDir("/a/b/c", 0755),
+	)
+
+	for _, tc := range []struct {
+		name   string
+		mounts func() ([]mount.Mount, error)
+	}{
+		{
+			name: "LoopOptions",
+			mounts: func() ([]mount.Mount, error) {
+				b1, err := initalizeBlockDevice(td, l1)
+				if err != nil {
+					return nil, err
+				}
+				b2, err := initalizeBlockDevice(td, l2)
+				if err != nil {
+					return nil, err
+				}
+				return []mount.Mount{
+					{
+						Type:    "xfs",
+						Source:  b1,
+						Options: []string{"loop"},
+					},
+					{
+						Type:    "xfs",
+						Source:  b2,
+						Options: []string{"loop"},
+					},
+					{
+						Type:   "format/overlay",
+						Source: "overlay",
+						Options: []string{
+							"lowerdir={{ mount 1 }}:{{ mount 0 }}",
+						},
+					},
+				}, nil
+			},
+		},
+		{
+			name: "SeparateLoop",
+			mounts: func() ([]mount.Mount, error) {
+				b1, err := initalizeBlockDevice(td, l1)
+				if err != nil {
+					return nil, err
+				}
+				b2, err := initalizeBlockDevice(td, l2)
+				if err != nil {
+					return nil, err
+				}
+				return []mount.Mount{
+					{
+						Type:    "loop",
+						Source:  b1,
+						Options: []string{},
+					},
+					{
+						Type:    "format/xfs",
+						Source:  "{{ mount 0 }}",
+						Options: []string{},
+					},
+					{
+						Type:    "loop",
+						Source:  b2,
+						Options: []string{},
+					},
+					{
+						Type:    "format/xfs",
+						Source:  "{{ mount 2 }}",
+						Options: []string{},
+					},
+					{
+						Type:   "format/overlay",
+						Source: "overlay",
+						Options: []string{
+							"lowerdir={{ mount 3 }}:{{ mount 1 }}",
+						},
+					},
+				}, nil
+			},
+		},
+		{
+			name: "OverlayFunc",
+			mounts: func() ([]mount.Mount, error) {
+				b1, err := initalizeBlockDevice(td, l1)
+				if err != nil {
+					return nil, err
+				}
+				b2, err := initalizeBlockDevice(td, l2)
+				if err != nil {
+					return nil, err
+				}
+				return []mount.Mount{
+					{
+						Type:    "xfs",
+						Source:  b1,
+						Options: []string{"loop"},
```

Comment:
one important thing I just have thought out is that
If one of the design goals is "to avoid snapshotters to mount filesystems theirselves", so that "mounts[] will be passed to mount handlers to parse".

If my understanding above is correct: I guess `loop` should not be decided by the snapshotter since (take EROFS for example):
 [runC] for linux versions without file-backed mounts: we should try to mount files with file-backed mounts, if -ENOTBLK is returned, use loopback devices;
 [runC] for linux versions with file-backed mounts: we should try to mount files with file-backed mounts, it should mount succeed;
 [secure containers like Kata]: we should use virtio-blk to pass through files rather than generating loops too;
 [gVisor]: they should mount files directly.

So I think the same file should be handled according to different shims, like runc-shim and kata-shim, etc.

And the file might be better to be parsed as:

[]mount.Mount{
   {
           Type: "erofs",
           Source: <fsmeta>
           Options: []string {
                "device=<layer0>,device=<layer1>,...,device=<layern-1>
           }
    },
<fsmeta>, <layer0>, <layer1>...<layern-1> are all files, it's up to mount managers to decide how to setup devices to mount these....